CONTRIBUTING.md

```diff
@@ -381,6 +381,68 @@ It offers a number of functionalities that improve development:
 
+## Unit tests
+
+The project has unit tests for both C++ and QML code.
+
+### C++ unit tests
+
+C++ unit tests are localized under `ui/StatusQ/tests`. They can be executed
+easily from QtCreator. Just open the `StatusQ` project in `QtCreator`
+(`status-desktop/ui/StatusQ/CMakeLists.txt`) and select the appropriate target (each test is a separate target).
+
+### QML unit tests
+
+QML tests are localized under `storybook/qmlTests/tests` and `ui/StatusQ/tests`. The
+first location is recommended for all new tests because it provides integration with
+the `Storybook`.
+
+Unit tests operate without backend. Backend layer is cut-off using stubs mechanism in the same
+way as for pages in `Storybook`.
+
+QML tests can be executed by running `QmlTests` target of `Storybook` project
+(setup described in section regarding `Storybook` above) or from `QtCreator` where only
+single test or subset of tests can be selected easily. Another convenient option is to
+use the test runner in `Storybook`, which gives ability to automatically execute tests
+for currently opened page when changes are detected in the code, or on demand using the
+button in the bottom bar. `Storybook` discovers tests by naming convention
+(e.g. `SwapModalPage.qml` will run `tst_SwapModal.qml` tests).
+
+### Unit tests recommendations
+
+- Unit tests should be independent and isolated.
+
+  They should not depend on anything except the unit of code under test. It means that
+  dependencies should be mocked and input data provided in the simplest possible form.
+  If the component under test needs a model, ideally it should be provided as a simple `ListModel`
+  instead of being created using additional components such as adapters or stores. Using an
+  additional component introduces a dependency on it in the test, even if the component under test
+  itself is independent of it.
+
+> [!NOTE]
+> Separately testing components that are used together in an application may lead to the
+> conclusion that it is uncertain whether they are compatible. However, the point is not to
+> not test it, but to do it in an appropriate test (usually in test of component using those
+> two components together).
+
+- Unit tests should test units.
+
+  This statement is also true regarding subcomponents used in a component tested in a given
+  unit test. Subcomponents should not be subject of testing in unit test (as they form
+  a different unit), they should be subject of testing in their own unit test. The unit tests
+  should verify if subcomponents are used in the intended way and simply assume that
+  internally they work as expected.
+
+  Respecting this rule will help keep unit tests smaller, better structured and easier
+  to maintain.
+
+- Unit tests should execute very quickly.
+
+  Relying on `wait(...)` calls slows down tests and is unreliable, may lead to flakey tests,
+  therefore should be avoided.
+
+- Unit tests should not generate any warnings and debug logs.
```

Comment:
I'm not sure what's the intent here. Is it referring that the test itself should not generate warnings and debug logs, or that there should be no warnings/debug logs generated by the component under test?

Totally agree with the first one, not sure about the second one 😄

Reply:
I think both. I added note about expected warnings from the component.

Reply:
I think it's good to have a goal of warning free tests, but probably not a rule.

It's quite difficult to achieve a warning free test because even some low level components will generate warnings. Sometimes it depends on Qt version. A good example here is the `StatusButton`. IIRC it shows a polish loop warning on some qt versions or platforms.